File: be/Services/NotificationService.cs
using FA23_Convocation2023_API.Models;
using FA23_Convocation2023_API.Hubs;
using Microsoft.EntityFrameworkCore;
using Microsoft.AspNetCore.SignalR;

namespace FA23_Convocation2023_API.Services
{
    public class NotificationService
    {
        private readonly Convo24Context _context;
        private readonly IHubContext<MessageHub> _hubContext;

        public NotificationService(Convo24Context context, IHubContext<MessageHub> hubContext)
        {
            _context = context;
            _hubContext = hubContext;
        }

        // Get all notifications with pagination
        public async Task<List<Notification>> GetAllNotificationsAsync(int pageIndex = 1, int pageSize = 10, string? status = null)
        {
            Console.WriteLine($"[DEBUG SERVICE] Getting notifications from database - pageIndex: {pageIndex}, pageSize: {pageSize}, status: {status}");

            // First, try to get total count to see if table has any data
            var totalCount = await _context.Notifications.CountAsync();
            Console.WriteLine($"[DEBUG SERVICE] Total notifications in database: {totalCount}");

            var query = _context.Notifications
                .AsQueryable();

            if (!string.IsNullOrEmpty(status))
            {
                Console.WriteLine($"[DEBUG SERVICE] Filtering by status: {status}");
                query = query.Where(n => n.Status == status);
            }

            var result = await query
                .OrderByDescending(n => n.CreatedAt)
                .Skip((pageIndex - 1) * pageSize)
                .Take(pageSize)
                .ToListAsync();

            Console.WriteLine($"[DEBUG SERVICE] Query returned {result.Count} notifications");
            return result;
        }

        // Get notification by ID
        public async Task<Notification?> GetNotificationByIdAsync(int id)
        {
            return await _context.Notifications
                .FirstOrDefaultAsync(n => n.NotificationId == id);
        }

        // Create new notification
        public async Task<Notification> CreateNotificationAsync(Notification notification)
        {
            notification.CreatedAt = DateTime.Now;
            notification.Status = "PENDING";

            _context.Notifications.Add(notification);
            await _context.SaveChangesAsync();

            // Broadcast new notification to NO group via SignalR
            await BroadcastNotificationToNoticers(notification, true);

            return notification;
        }

        // Helper method to broadcast notification to NO group
        private async Task BroadcastNotificationToNoticers(Notification notification, bool isNewNotification = false)
        {
            Console.WriteLine($"[SERVICE] Broadcasting notification {notification.NotificationId} to NO group");

            // Get hall and session info if available
            var hall = notification.HallId.HasValue ? await _context.Halls.FindAsync(notification.HallId.Value) : null;
            var session = notification.SessionId.HasValue ? await _context.Sessions.FindAsync(notification.SessionId.Value) : null;

            var broadcastData = new
            {
                NotificationId = notification.NotificationId,
                Title = notification.Title,
                Content = notification.Content,
                Priority = notification.Priority,
                PriorityText = GetPriorityText(notification.Priority),
                RepeatCount = notification.RepeatCount,
                HallName = hall?.HallName,
                SessionNumber = session?.Session1,
                Scope = GetNotificationScope(notification, hall, session),
                BroadcastAt = DateTime.UtcNow,
                IsNewNotification = isNewNotification // Flag to indicate if this is a new notification
            };

            Console.WriteLine($"[SERVICE DEBUG] Broadcasting data: {System.Text.Json.JsonSerializer.Serialize(broadcastData)}");
            await _hubContext.Clients.Group("NO").SendAsync("ReceiveTTSBroadcast", broadcastData);
            Console.WriteLine($"[SERVICE] Notification {notification.NotificationId} broadcasted to NO group successfully");
        }

        // Helper method to get priority text
        private string GetPriorityText(int priority)
        {
            return priority switch
            {
                1 => "High",
                2 => "Medium",
                3 => "Low",
                _ => "Normal"
            };
        }

        // Helper method to get notification scope
        private string GetNotificationScope(Notification notification, Hall? hall, Session? session)
        {
            if (notification.HallId.HasValue && notification.SessionId.HasValue)
            {
                return $"Hội trường {hall?.HallName} - Đợt {session?.Session1}";
            }
            else if (notification.HallId.HasValue)
            {
                return $"Hội trường {hall?.HallName}";
            }
            else if (notification.SessionId.HasValue)
            {
                return $"Đợt {session?.Session1}";
            }
            else
            {
                return "Toàn trường";
            }
        }

        // Update notification
        public async Task<Notification?> UpdateNotificationAsync(int id, Notification updatedNotification)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null) return null;

            notification.Title = updatedNotification.Title;
            notification.Content = updatedNotification.Content;
            notification.Priority = updatedNotification.Priority;
            notification.HallId = updatedNotification.HallId;
            notification.SessionId = updatedNotification.SessionId;
            notification.ScheduledAt = updatedNotification.ScheduledAt;
            notification.IsAutomatic = updatedNotification.IsAutomatic;
            notification.RepeatCount = updatedNotification.RepeatCount;

            await _context.SaveChangesAsync();
            return notification;
        }

        // Delete notification
        public async Task<bool> DeleteNotificationAsync(int id)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null) return false;

            _context.Notifications.Remove(notification);
            await _context.SaveChangesAsync();
            return true;
        }

        // Mark notification as broadcasting
        public async Task<bool> StartBroadcastAsync(int id, string broadcastByUserId)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null || notification.Status != "PENDING") return false;

            notification.Status = "BROADCASTING";
            notification.BroadcastBy = broadcastByUserId;
            notification.BroadcastAt = DateTime.Now;

            await _context.SaveChangesAsync();
            return true;
        }

        // Mark notification as completed
        public async Task<bool> CompleteBroadcastAsync(int id)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null || notification.Status != "BROADCASTING") return false;

            notification.Status = "COMPLETED";
            await _context.SaveChangesAsync();
            return true;
        }

        // Cancel notification
        public async Task<bool> CancelNotificationAsync(int id)
        {
            var notification = await _context.Notifications.FindAsync(id);
            if (notification == null || notification.Status == "COMPLETED") return false;

            notification.Status = "CANCELLED";
            await _context.SaveChangesAsync();
            return true;
        }

        // Get pending notifications for auto broadcast
        public async Task<List<Notification>> GetPendingAutomaticNotificationsAsync()
        {
            return await _context.Notifications
                .Where(n => n.Status == "PENDING" &&
                           n.IsAutomatic &&
                           n.ScheduledAt != null &&
                           n.ScheduledAt <= DateTime.Now)
                .Include(n => n.Hall)
                .Include(n => n.Session)
                .ToListAsync();
        }

        // Get notifications by hall and session
        public async Task<List<Notification>> GetNotificationsByHallAndSessionAsync(int? hallId, int? sessionId)
        {
            var query = _context.Notifications.AsQueryable();

            if (hallId.HasValue)
            {
                query = query.Where(n => n.HallId == hallId || n.HallId == null);
            }

            if (sessionId.HasValue)
            {
                query = query.Where(n => n.SessionId == sessionId || n.SessionId == null);
            }

            return await query
                .Where(n => n.Status == "PENDING")
                .OrderBy(n => n.Priority)
                .ThenBy(n => n.CreatedAt)
                .ToListAsync();
        }
    }
}

File: be-nest/src/notification/notification.service.ts
import { Inject, Injectable } from '@nestjs/common';
import { and, asc, count, desc, eq, lte, or, sql, type SQL } from 'drizzle-orm';
import { alias } from 'drizzle-orm/pg-core';
import { DATABASE } from '../database/database.constants.js';
import type { AppDatabase } from '../database/database.types.js';
import { authUser } from '../database/schema/auth-schema.js';
import { halls, notifications, sessions } from '../database/schema/domain-schema.js';
import { RealtimeService } from '../realtime/realtime.service.js';
import type { TtsBroadcast } from '../realtime/realtime-events.js';
import type { NotificationDto } from './dto/notification.dto.js';

const createdByUser = alias(authUser, 'created_by_user');
const broadcastByUser = alias(authUser, 'broadcast_by_user');

@Injectable()
export class NotificationService {
  constructor(
    @Inject(DATABASE) private readonly database: AppDatabase,
    private readonly realtime: RealtimeService,
  ) {}

  async list(pageIndex: number, pageSize: number, status?: string) {
    const condition = status ? eq(notifications.status, status) : undefined;
    const rows = await this.responseQuery(condition)
      .orderBy(desc(notifications.createdAt))
      .limit(pageSize)
      .offset((pageIndex - 1) * pageSize);
    const total = (
      await this.database.select({ count: count() }).from(notifications).where(condition)
    )[0]?.count ?? 0;
    return {
      notifications: rows.map((row) => this.toResponse(row)),
      totalCount: total,
      pageIndex,
      pageSize,
      totalPages: Math.ceil(total / pageSize),
    };
  }

  async get(id: number) {
    const [row] = await this.responseQuery(eq(notifications.id, id)).limit(1);
    return row ? this.toResponse(row) : null;
  }

  async create(input: NotificationDto, actorId: string) {
    const [created] = await this.database
      .insert(notifications)
      .values({ ...input, createdBy: actorId, status: 'PENDING' })
      .returning();
    const payload = await this.broadcastPayload(created!.id, true);
    if (payload) this.realtime.broadcastTts(payload);
    this.realtime.notificationChanged(created!.id, created!.status);
    return created!;
  }

  async update(id: number, input: NotificationDto): Promise<boolean> {
    const rows = await this.database
      .update(notifications)
      .set(input)
      .where(eq(notifications.id, id))
      .returning();
    if (rows.length) this.realtime.notificationChanged(id, rows[0]!.status);
    return rows.length > 0;
  }

  async delete(id: number): Promise<boolean> {
    return (await this.database.delete(notifications).where(eq(notifications.id, id)).returning())
      .length > 0;
  }

  async transition(id: number, from: string[], status: string, actorId?: string): Promise<boolean> {
    const conditions = [eq(notifications.id, id), or(...from.map((value) => eq(notifications.status, value)))!];
    const rows = await this.database
      .update(notifications)
      .set({
        status,
        ...(status === 'BROADCASTING'
          ? { broadcastBy: actorId, broadcastAt: new Date() }
          : {}),
      })
      .where(and(...conditions))
      .returning();
    if (rows.length) this.realtime.notificationChanged(id, status);
    return rows.length > 0;
  }

  async pending(hallId?: number, sessionId?: number) {
    const filters: SQL[] = [eq(notifications.status, 'PENDING')];
    if (hallId !== undefined) filters.push(or(eq(notifications.hallId, hallId), sql`${notifications.hallId} is null`)!);
    if (sessionId !== undefined) filters.push(or(eq(notifications.sessionId, sessionId), sql`${notifications.sessionId} is null`)!);
    const rows = await this.responseQuery(and(...filters))
      .orderBy(asc(notifications.priority), asc(notifications.createdAt));
    return rows.map((row) => this.toResponse(row));
  }

  async pendingAutomatic() {
    return this.database
      .select()
      .from(notifications)
      .where(
        and(
          eq(notifications.status, 'PENDING'),
          eq(notifications.isAutomatic, true),
          lte(notifications.scheduledAt, new Date()),
        ),
      );
  }

  async broadcast(id: number): Promise<TtsBroadcast | null> {
    const payload = await this.broadcastPayload(id, false);
    if (payload) this.realtime.broadcastTts(payload);
    return payload;
  }

  private responseQuery(condition?: SQL) {
    return this.database
      .select({
        notification: notifications,
        hallName: halls.name,
        sessionNumber: sessions.sessionNumber,
        createdByName: createdByUser.name,
        broadcastByName: broadcastByUser.name,
      })
      .from(notifications)
      .leftJoin(halls, eq(halls.id, notifications.hallId))
      .leftJoin(sessions, eq(sessions.id, notifications.sessionId))
      .leftJoin(createdByUser, eq(createdByUser.id, notifications.createdBy))
      .leftJoin(broadcastByUser, eq(broadcastByUser.id, notifications.broadcastBy))
      .where(condition);
  }

  private async broadcastPayload(id: number, isNewNotification: boolean): Promise<TtsBroadcast | null> {
    const notification = await this.get(id);
    if (!notification) return null;
    return {
      notificationId: notification.notificationId,
      title: notification.title,
      content: notification.content,
      priority: notification.priority,
      priorityText: notification.priorityText,
      repeatCount: notification.repeatCount,
      hallName: notification.hallName,
      sessionNumber: notification.sessionNumber,
      scope: notification.scope,
      broadcastAt: new Date().toISOString(),
      isNewNotification,
    };
  }

  private toResponse(row: Awaited<ReturnType<NotificationService['responseQuery']>>[number]) {
    const value = row.notification;
    return {
      notificationId: value.id,
      title: value.title,
      content: value.content,
      priority: value.priority,
      priorityText: this.priorityText(value.priority),
      hallId: value.hallId,
      hallName: row.hallName,
      sessionId: value.sessionId,
      sessionNumber: row.sessionNumber,
      createdBy: value.createdBy,
      createdByName: row.createdByName ?? '',
      broadcastBy: value.broadcastBy,
      broadcastByName: row.broadcastByName,
      createdAt: value.createdAt,
      scheduledAt: value.scheduledAt,
      broadcastAt: value.broadcastAt,
      status: value.status,
      isAutomatic: value.isAutomatic,
      repeatCount: value.repeatCount,
      scope: this.scope(row.hallName, row.sessionNumber),
    };
  }

  private priorityText(priority: number): string {
    return priority === 1 ? 'High' : priority === 2 ? 'Medium' : priority === 3 ? 'Low' : 'Normal';
  }

  private scope(hallName: string | null, sessionNumber: number | null): string {
    if (hallName && sessionNumber !== null) return `Hội trường ${hallName} - Đợt ${sessionNumber}`;
    if (hallName) return `Hội trường ${hallName}`;
    if (sessionNumber !== null) return `Đợt ${sessionNumber}`;
    return 'Toàn trường';
  }
}
